### ISSUE TITLE
SNV only added on forward read when both read ends overlap the position

### ISSUE THREAD

**Issue by rebber** [adamewing/bamsurgeon#132]
Hi,

I want to use bamsurgeon for resetting specific SNPs to homozygous reference in my sample. 

I have run addsnv.py (latest available version) as:
```
addsnv.py -v $var -f $inbam -r $bwaref -o $outbam -p $threads \
    --picardjar ~/miniconda2/share/picard-2.14-0/picard.jar --tmpdir $tmp \
    --tagreads --aligner mem --alignopts M:1 --ignoresnps --ignoreref -z 146
```

I used options --ignoresnps and --ignoreref, to enable setting SNPs to reference, and -z 146 to allow this even if they appear on the same read. 

However, for 31 of 100 SNPs I still get a heterozygous call when running HaplotypeCaller on the output bam of addsnv.py. I checked the mutated bam in these positions in IGV, and found that most of the reads still containing the SNP were reverse reads where the forward read also overlapped the SNP and only the forward read was mutated (to reference). 

Commit 894d6dae11e6a73c574260e71f5868a30c05cc8f seem to be claiming a fix to this, but I don't understand really what's happening in that part of the code.

Is this the expected behaviour of bamsurgeon when the two ends of a read pair both overlap the mutation position, i.e. only the forward read being mutated? Is there a way to avoid this, and have both ends mutated?

Best regards
Rebecka

**Comment by adamewing** [comment]
Hi Rebecka,

Having both ends of a read overlapping the mutation site _should_ work but perhaps you've found a situation where it doesn't. If you leave the `-z 146` option off does that help? Would you be able to send me a segment of the input .bam file containing a site where you have this problem? (please only do so if it's not patient data or otherwise sensitive).

Best,
Adam

**Comment by rebber** [comment]
Hi Adam,

I have tried without the -z option as well, and the results are the same.
My data is from an anonymized healthy donor, so I should be able to share it. I have a bam file with reads covering some of the SNPs in a google drive folder, which I will share with you. It both contains SNPs for which bamsurgeon was successful in resetting to reference and SNPs for which bamsurgeon failed this, mainly due to the issue with overlapping reads (defined as giving a heterozygous call by HaplotypeCaller). The folder also contains two variant files, one listing the successful resets/mutations and one listing the failed resets/mutations. 

It would be great if you could have a look at this and see if there's any issue I couldn't find.

Best regards 
Rebecka

**Comment by adamewing** [comment]
Hi Rebecka,

Thanks for your help. What I'm seeing for the failed resets is 1-2% of reads aren't converted back to reference which is probably what's triggering the het calls. This appeared to be largely driven by cases where the target base was the last one in the read. You also need to ensure the input bam files doesn't have unpaired reads as these will be ignored by addsnv (probably not an issue on your end). Try pulling the latest commit (d611518) and re-running, let me know if the results look better.

--Adam

**Comment by rebber** [comment]
Hi Adam,

Thank for looking in to this. I've pulled and installed the latest commits, but still have the same problem. I only get a marginally better result when running HaplotypeCaller on the mutated bam with the updated bamsurgeon. Now 28/100 SNPs instead of 31/100 SNPs are still called as heterozygous reference by HaplotypeCaller. I still have the problem of the reverse read not being mutated when both reads in a pair are overlapping the position of interest. This also happens in the postions which are not called by HaplotypeCaller. My guess is that the fraction of overlapping reads is less in positions not called, or that differences in base qualities make HaplotypeCaller prefer the forward read. Anyway, the question here is why does bamsurgeon not mutate the reverse reads? 

I have studied some of the positions in IGV, and can't find anything weird with the reads other than that they are overlapping. Here is an example for a read pair marked in blue.

The first image shows the reverse read, which has the SNP (i.e. bamsurgeon was not abled to mutate this position and read):
![Screen Shot 2019-07-01 at 13 25 45](https://user-images.githubusercontent.com/16937395/60439422-23c5fb80-9c13-11e9-8a1c-ebb27436585a.png)

The second image shows the forward read in the same pair, which is mutated to reference:
![Screen Shot 2019-07-01 at 13 28 39](https://user-images.githubusercontent.com/16937395/60439514-4e17b900-9c13-11e9-8a19-cb468221ce1c.png)

The third image shows the whole pair with "View as pairs", where you clearly see that the reads are overlapping at the position of interest:
![Screen Shot 2019-07-01 at 13 29 42](https://user-images.githubusercontent.com/16937395/60439597-612a8900-9c13-11e9-843f-327b6c7de1da.png)

Mapping quality in both ends are good, and so is base quality. There is nothing suspicious to find from the flags or tags, and the position of interest is not particularly close to an end in this read. This pattern is the same for the vast majority of the reads where bamsurgeon was not succesful.

I've also uploaded new bam file and positions called and not called by HaplotypeCaller after running with the latest version, to the same Google Drive folder as before.

Could you please have a second look at the cause of this? I'll be on vacation until end of July now though, so there's no immediate hurry for me.

Best regards
Rebecka

**Comment by rbrns** [comment]
I'm having a similar issue where only the forward read is edited when there are overlapping read-pairs.  I'm using the latest code available.
![bamsurgeon_snapshot](https://user-images.githubusercontent.com/37840424/61799401-af850f00-adf0-11e9-8766-e25f7164ad7b.png)
The top pane shows the unedited bam while the bottom pane shows the bam surgeon result.  Here I am attempting to change all bases at the highlighted locus to T.

I am running bam surgeon as:
`addsnv.py --varfile snps_for_bamsurgeon.txt --bamfile Un-edited.bam --reference ~/reference/hg19/ucsc.hg19.fasta  --outbam BamSurgeon_edited.bam --force --aligner mem --picardjar ~/miniconda3/envs/bamsurgeon/share/picard-2.20.4-0/picard.jar`

I've also experimented with --ignoresnps, however, that did not improve things.

This is amplicon data, so there can be an abundance of overlapping read pairs.

**Comment by adamewing** [comment]
Hopefully fixed by @darichter87 in 20d431e.
